### PR TITLE
Add compatibility for ibex

### PR DESCRIPTION
**Pull request by micprog**
The request adds compatibility for ibex as the core used in pulp. Following modifications were necessary:

- limit instructions to RV32IMC (no compatibility for RI5CY XPULPV2 instructions, no `__builtin_*` functions)
- Use custom interrupts (ibex added special registers for pulp compatibility -> `SR_MTVEC = 0x7D1`)
- Add ibex performance counters (see https://ibex-core.readthedocs.io/en/latest/03_reference/performance_counters.html)

To use, set environment variable `USE_IBEX`.

### CODE REVIEW

File: include/archi/ibex/mhpm.h [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+#define IBEX_NumCyclesLSU     3  /* Number of cycles waiting for data memory */
+#define IBEX_NumCyclesIF      4  /* Cycles waiting for instruction fetches, i.e., number of instructions wasted due to non-ideal caching */
+#define IBEX_NumLoads         5  /* Number of data memory loads. Misaligned accesses are counted as two accesses */
+#define IBEX_NumStores        6  /* Number of data memory stores. Misaligned accesses are counted as two accesses */
+#define IBEX_NumJumps         7  /* Number of unconditional jumps (j, jal, jr, jalr) */
+#define IBEX_NumBranches      8  /* Number of branches (conditional) */
+#define IBEX_NumBranchesTaken 9  /* Number of taken branches (conditional) */
+#define IBEX_NumInstrRetC     10 /* Number of compressed instructions retired */
+#define IBEX_NumCyclesMulWait 11 /* Cycles waiting for multiply to complete */
+#define IBEX_NumCyclesDivWait 12 /* Cycles waiting for divide to complete */
```
Comment (bluewww) [inline include/archi/ibex/mhpm.h:40]: Minor nit: Macros should be all uppercase

File: include/archi/ibex/mhpm.h [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+#define CSR_PCER_BRANCH       IBEX_NumBranches      /* Number of branch instructions seen, i.e. bf, bnf */
+#define CSR_PCER_TAKEN_BRANCH IBEX_NumBranchesTaken /* Number of taken branch instructions seen, i.e. bf, bnf */
+#define CSR_PCER_RVC          IBEX_NumInstrRetC     /* Number of compressed instructions */
+// Not implemented in ibex
+// #define CSR_PCER_ELW          0   /* Cycles wasted due to ELW instruction */
+// #define CSR_PCER_JMP_STALL    0   /* Number of jump register hazards */
+
+
+// External perf counters
+#define IBEX_External_BASE    13
```
Comment (bluewww) [inline include/archi/ibex/mhpm.h:60]: dito

File: include/hal/ibex/ibex.h [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+// For PULP, ibex added non-standard irqs to allow for 32 fast interrupts. 
+// These use custom CSRs: MIE: 0x7D0, MTVEC: 0x7D1, MIP: 0x7D2
+#define SR_MTVEC  0x7D1
+
+#define hal_spr_read_then_clr(reg,val) \
+  ({ \
+    int state; \
+    asm volatile ("csrrc %0, %1, %2" :  "=r" (state) : "I" (reg), "I" (val) ); \
+    state; \
+  })
```
Comment (bluewww) [inline include/hal/ibex/ibex.h:33]: I guess at some point I will make this file the global default.

There is literally zero reason why we have builtins to read csr registers I don't know who thought this is a good idea.

Anyway this is good stuff

**Review by bluewww — CHANGES_REQUESTED** [review]
Just change the macros and its good to go.

If you like you could add also a configuration script in `configs/` for the pulp + ibex configuration. All you need to do there is just copy copy paste `configs/pulp.sh`t to something like `configs/pulp-ibex.sh` and add a `export USE_IBEX=1` line.

File: include/archi/ibex/mhpm.h (hunk 1 above)
Comment (micprog) [reply include/archi/ibex/mhpm.h:40]: I used the names from the [ibex docs](https://ibex-core.readthedocs.io/en/latest/03_reference/performance_counters.html), but I will update this.

**Review by bluewww — APPROVED** [review]
I briefly tested hello world on pulpissimo + ri5cy and pulpissimo + ibex. Seems to be ok.
lgtm.

**Pull request merged**